File: scripts/create-snapshot.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

/**
 * Создает снапшот указанной папки со всеми файлами и их содержимым
 * Usage: node scripts/create-snapshot.js [folder] [output]
 * Example: node scripts/create-snapshot.js src src-snapshot.md
 */

// Конфигурация
const CONFIG = {
  // Расширения файлов для включения в снапшот
  includeExtensions: [
    '.ts', '.tsx', '.js', '.jsx',
    '.css', '.scss', '.sass', '.less',
    '.json', '.md', '.txt', '.yml', '.yaml',
    '.html', '.xml', '.svg',
    '.env', '.gitignore', '.eslintrc',
    // Добавьте другие расширения по необходимости
  ],
  
  // Папки и файлы для исключения
  excludePatterns: [
    'node_modules',
    '.git',
    '.next',
    'dist',
    'build',
    '.turbo',
    'coverage',
    '.nyc_output',
    'logs',
    '*.log',
    '.DS_Store',
    'Thumbs.db',
    '*.tmp',
    '*.temp',
  ],
  
  // Максимальный размер файла в байтах (100KB)
  maxFileSize: 100 * 1024,
  
  // Максимальная глубина вложенности
  maxDepth: 10,
};

/**
 * Проверяет, следует ли исключить файл/папку
 */
function shouldExclude(name, isDirectory = false) {
  for (const pattern of CONFIG.excludePatterns) {
    if (pattern.includes('*')) {
      // Простая поддержка wildcards
      const regex = new RegExp(pattern.replace(/\*/g, '.*'));
      if (regex.test(name)) return true;
    } else {
      if (name === pattern) return true;
    }
  }
  return false;
}

/**
 * Проверяет, следует ли включить файл
 */
function shouldIncludeFile(filename) {
  const ext = path.extname(filename).toLowerCase();
  
  // Файлы без расширения (как Dockerfile, README)
  if (!ext && /^[A-Z][A-Z_]*[A-Z]?$/.test(path.basename(filename))) {
    return true;
  }
  
  return CONFIG.includeExtensions.includes(ext);
}

/**
 * Получает язык для подсветки синтаксиса по расширению файла
 */
function getLanguageByExtension(filename) {
  const ext = path.extname(filename).toLowerCase();
  const basename = path.basename(filename).toLowerCase();
  
  const langMap = {
    '.ts': 'typescript',
    '.tsx': 'tsx',
    '.js': 'javascript',
    '.jsx': 'jsx',
    '.css': 'css',
    '.scss': 'scss',
    '.sass': 'sass',
    '.less': 'less',
    '.json': 'json',
    '.md': 'markdown',
    '.html': 'html',
    '.xml': 'xml',
    '.yml': 'yaml',
    '.yaml': 'yaml',
    '.svg': 'xml',
    '.txt': 'text',
  };
  
  // Специальные случаи для файлов без расширения
  if (!ext) {
    if (basename === 'dockerfile') return 'dockerfile';
    if (basename === 'makefile') return 'makefile';
    if (basename.includes('gitignore')) return 'gitignore';
    if (basename.includes('eslintrc')) return 'json';
    return 'text';
  }
  
  return langMap[ext] || 'text';
}

/**
 * Создает дерево файловой структуры
 */
function createFileTree(dirPath, prefix = '', isLast = true, depth = 0) {
  if (depth > CONFIG.maxDepth) return '';
  
  const name = path.basename(dirPath);
  if (shouldExclude(name, true)) return '';
  
  let tree = '';
  const connector = isLast ? '└── ' : '├── ';
  tree += `${prefix}${connector}${name}/\n`;
  
  try {
    const items = fs.readdirSync(dirPath, { withFileTypes: true });
    const filteredItems = items.filter(item => {
      if (shouldExclude(item.name, item.isDirectory())) return false;
      if (item.isFile() && !shouldIncludeFile(item.name)) return false;
      return true;
    });
    
    filteredItems.forEach((item, index) => {
      const isLastItem = index === filteredItems.length - 1;
      const newPrefix = prefix + (isLast ? '    ' : '│   ');
      const itemPath = path.join(dirPath, item.name);
      
      if (item.isDirectory()) {
        tree += createFileTree(itemPath, newPrefix, isLastItem, depth + 1);
      } else {
        const fileConnector = isLastItem ? '└── ' : '├── ';
        tree += `${newPrefix}${fileConnector}${item.name}\n`;
      }
    });
  } catch (error) {
    console.warn(`Не удалось прочитать директорию ${dirPath}: ${error.message}`);
  }
  
  return tree;
}

/**
 * Собирает все файлы рекурсивно
 */
function collectFiles(dirPath, depth = 0) {
  if (depth > CONFIG.maxDepth) return [];
  
  const files = [];
  
  try {
    const items = fs.readdirSync(dirPath, { withFileTypes: true });
    
    for (const item of items) {
      if (shouldExclude(item.name, item.isDirectory())) continue;
      
      const itemPath = path.join(dirPath, item.name);
      
      if (item.isDirectory()) {
        files.push(...collectFiles(itemPath, depth + 1));
      } else if (shouldIncludeFile(item.name)) {
        try {
          const stats = fs.statSync(itemPath);
          if (stats.size <= CONFIG.maxFileSize) {
            files.push(itemPath);
          } else {
            console.warn(`Файл ${itemPath} пропущен (размер ${stats.size} байт > ${CONFIG.maxFileSize})`);
          }
        } catch (error) {
          console.warn(`Не удалось получить информацию о файле ${itemPath}: ${error.message}`);
        }
      }
    }
  } catch (error) {
    console.warn(`Не удалось прочитать директорию ${dirPath}: ${error.message}`);
  }
  
  return files;
}

/**
 * Создает снапшот
 */
function createSnapshot(sourceDir, outputFile) {
  const absoluteSourceDir = path.resolve(sourceDir);
  const sourceName = path.basename(absoluteSourceDir);
  
  if (!fs.existsSync(absoluteSourceDir)) {
    console.error(`❌ Папка ${sourceDir} не существует`);
    process.exit(1);
  }
  
  // Создаем папку snapshots если не существует
  const snapshotsDir = 'snapshots';
  if (!fs.existsSync(snapshotsDir)) {
    fs.mkdirSync(snapshotsDir, { recursive: true });
    console.log(`📁 Создана папка: ${snapshotsDir}`);
  }
  
  // Корректируем путь к выходному файлу
  const finalOutputFile = path.join(snapshotsDir, path.basename(outputFile));
  
  console.log(`📁 Создание снапшота папки: ${sourceDir}`);
  console.log(`📄 Выходной файл: ${finalOutputFile}`);
  
  let markdown = `# Снапшот папки ${sourceName}\n\n`;
  markdown += `*Создано: ${new Date().toLocaleString('ru-RU')}*\n\n`;
  
  // Добавляем дерево структуры
  console.log('🌳 Генерация дерева файлов...');
  markdown += '## Структура файлов\n\n```\n';
  markdown += `${sourceName}/\n`;
  
  try {
    const items = fs.readdirSync(absoluteSourceDir, { withFileTypes: true });
    const filteredItems = items.filter(item => {
      if (shouldExclude(item.name, item.isDirectory())) return false;
      if (item.isFile() && !shouldIncludeFile(item.name)) return false;
      return true;
    });
    
    filteredItems.forEach((item, index) => {
      const isLast = index === filteredItems.length - 1;
      const itemPath = path.join(absoluteSourceDir, item.name);
      
      if (item.isDirectory()) {
        markdown += createFileTree(itemPath, '', isLast, 1);
      } else {
        const connector = isLast ? '└── ' : '├── ';
        markdown += `${connector}${item.name}\n`;
      }
    });
  } catch (error) {
    console.error(`❌ Ошибка при создании дерева: ${error.message}`);
  }
  
  markdown += '```\n\n---\n\n';
  
  // Собираем все файлы
  console.log('📋 Сбор файлов...');
  const files = collectFiles(absoluteSourceDir);
  console.log(`📊 Найдено файлов: ${files.length}`);
  
  if (files.length === 0) {
    console.warn('⚠️  Файлы для включения в снапшот не найдены');
    return;
  }
  
  // Добавляем содержимое файлов
  markdown += '## Содержимое файлов\n\n';
  
  files.sort().forEach((filePath, index) => {
    const relativePath = path.relative(absoluteSourceDir, filePath);
    const normalizedPath = relativePath.replace(/\\/g, '/'); // Для Windows
    const filename = path.basename(filePath);
    
    console.log(`📄 Обработка [${index + 1}/${files.length}]: ${normalizedPath}`);
    
    try {
      const content = fs.readFileSync(filePath, 'utf8');
      const language = getLanguageByExtension(filename);
      
      markdown += `### ${normalizedPath}\n`;
      markdown += `\`\`\`${language}\n`;
      markdown += content;
      markdown += '\n```\n\n';
    } catch (error) {
      console.warn(`⚠️  Не удалось прочитать файл ${filePath}: ${error.message}`);
      markdown += `### ${normalizedPath}\n`;
      markdown += `*Ошибка чтения файла: ${error.message}*\n\n`;
    }
  });
  
  // Записываем результат
  try {
    fs.writeFileSync(finalOutputFile, markdown, 'utf8');
    console.log(`✅ Снапшот создан: ${finalOutputFile}`);
    
    const stats = fs.statSync(finalOutputFile);
    const sizeKB = Math.round(stats.size / 1024);
    console.log(`📊 Размер файла: ${sizeKB} KB`);
  } catch (error) {
    console.error(`❌ Ошибка записи файла: ${error.message}`);
    process.exit(1);
  }
}

/**
 * Показывает помощь
 */
function showHelp() {
  console.log(`
📸 Скрипт создания снапшотов папок

Использование:
  node scripts/create-snapshot.js [папка] [выходной_файл]

Примеры:
  node scripts/create-snapshot.js src src-snapshot.md
  node scripts/create-snapshot.js components components-snapshot.md
  node scripts/create-snapshot.js . full-project-snapshot.md

Опции:
  -h, --help     Показать эту справку

Настройки (в файле scripts/create-snapshot.js):
  - Поддерживаемые расширения: ${CONFIG.includeExtensions.join(', ')}
  - Исключаемые папки: ${CONFIG.excludePatterns.join(', ')}
  - Максимальный размер файла: ${Math.round(CONFIG.maxFileSize / 1024)} KB
  - Максимальная глубина: ${CONFIG.maxDepth} уровней
`);
}

// Главная функция
function main() {
  const args = process.argv.slice(2);
  
  if (args.includes('-h') || args.includes('--help')) {
    showHelp();
    process.exit(0);
  }
  
  const sourceDir = args[0] || 'src';
  const outputFile = args[1] || `${path.basename(sourceDir)}-snapshot.md`;
  
  console.log('📸 Создание снапшота...');
  createSnapshot(sourceDir, outputFile);
}

// Запуск если файл выполняется напрямую
if (require.main === module) {
  main();
}

module.exports = { createSnapshot, CONFIG };

File: scripts/create-snapshot.py
#!/usr/bin/env python3
"""
Создает снапшот указанной папки со всеми файлами и их содержимым
Usage: python scripts/create-snapshot.py [folder] [output]
Example: python scripts/create-snapshot.py src src-snapshot.md
"""

import os
import sys
from pathlib import Path
from datetime import datetime
import argparse

# Конфигурация
CONFIG = {
    # Расширения файлов для включения в снапшот
    'include_extensions': {
        '.ts', '.tsx', '.js', '.jsx',
        '.css', '.scss', '.sass', '.less',
        '.json', '.md', '.txt', '.yml', '.yaml',
        '.html', '.xml', '.svg',
        '.env', '.gitignore', '.eslintrc', '.py'
    },
    
    # Папки и файлы для исключения
    'exclude_patterns': {
        'node_modules', '.git', '.next', 'dist', 'build',
        '.turbo', 'coverage', '.nyc_output', 'logs',
        '__pycache__', '.pytest_cache', 'venv', 'env',
        '.venv', '.DS_Store', 'Thumbs.db'
    },
    
    # Максимальный размер файла в байтах (100KB)
    'max_file_size': 100 * 1024,
    
    # Максимальная глубина вложенности
    'max_depth': 10,
}

def get_language_by_extension(filename):
    """Получает язык для подсветки синтаксиса по расширению файла"""
    ext = Path(filename).suffix.lower()
    basename = Path(filename).name.lower()
    
    lang_map = {
        '.ts': 'typescript',
        '.tsx': 'tsx',
        '.js': 'javascript', 
        '.jsx': 'jsx',
        '.py': 'python',
        '.css': 'css',
        '.scss': 'scss',
        '.sass': 'sass',
        '.less': 'less',
        '.json': 'json',
        '.md': 'markdown',
        '.html': 'html',
        '.xml': 'xml',
        '.yml': 'yaml',
        '.yaml': 'yaml',
        '.svg': 'xml',
        '.txt': 'text',
    }
    
    # Специальные случаи для файлов без расширения
    if not ext:
        if 'dockerfile' in basename:
            return 'dockerfile'
        elif 'makefile' in basename:
            return 'makefile'
        elif 'gitignore' in basename:
            return 'gitignore'
        elif 'eslintrc' in basename:
            return 'json'
        return 'text'
    
    return lang_map.get(ext, 'text')

def should_exclude(name, is_directory=False):
    """Проверяет, следует ли исключить файл/папку"""
    return name in CONFIG['exclude_patterns'] or name.startswith('.')

def should_include_file(filename):
    """Проверяет, следует ли включить файл"""
    path = Path(filename)
    ext = path.suffix.lower()
    
    # Файлы без расширения (как Dockerfile, README)
    if not ext and path.name.isupper():
        return True
    
    return ext in CONFIG['include_extensions']

def create_file_tree(dir_path, prefix='', is_last=True, depth=0):
    """Создает дерево файловой структуры"""
    if depth > CONFIG['max_depth']:
        return ''
    
    name = Path(dir_path).name
    if should_exclude(name, True):
        return ''
    
    tree = ''
    connector = '└── ' if is_last else '├── '
    tree += f"{prefix}{connector}{name}/\n"
    
    try:
        items = []
        for item in Path(dir_path).iterdir():
            if should_exclude(item.name, item.is_dir()):
                continue
            if item.is_file() and not should_include_file(item.name):
                continue
            items.append(item)
        
        items.sort(key=lambda x: (x.is_file(), x.name.lower()))
        
        for i, item in enumerate(items):
            is_last_item = i == len(items) - 1
            new_prefix = prefix + ('    ' if is_last else '│   ')
            
            if item.is_dir():
                tree += create_file_tree(item, new_prefix, is_last_item, depth + 1)
            else:
                file_connector = '└── ' if is_last_item else '├── '
                tree += f"{new_prefix}{file_connector}{item.name}\n"
                
    except PermissionError:
        print(f"⚠️  Нет доступа к директории {dir_path}")
    except Exception as e:
        print(f"⚠️  Ошибка при чтении директории {dir_path}: {e}")
    
    return tree

def collect_files(dir_path, depth=0):
    """Собирает все файлы рекурсивно"""
    if depth > CONFIG['max_depth']:
        return []
    
    files = []
    try:
        for item in Path(dir_path).rglob('*'):
            if item.is_file():
                # Проверяем путь на исключения
                parts = item.parts
                if any(should_exclude(part, True) for part in parts):
                    continue
                
                if not should_include_file(item.name):
                    continue
                
                try:
                    if item.stat().st_size <= CONFIG['max_file_size']:
                        files.append(item)
                    else:
                        size_kb = item.stat().st_size // 1024
                        print(f"⚠️  Файл {item} пропущен (размер {size_kb} KB)")
                except Exception as e:
                    print(f"⚠️  Не удалось получить информацию о файле {item}: {e}")
                    
    except Exception as e:
        print(f"⚠️  Ошибка при сборе файлов из {dir_path}: {e}")
    
    return files

def create_snapshot(source_dir, output_file):
    """Создает снапшот"""
    source_path = Path(source_dir).resolve()
    source_name = source_path.name
    
    if not source_path.exists():
        print(f"❌ Папка {source_dir} не существует")
        sys.exit(1)
    
    # Создаем папку snapshots если не существует
    snapshots_dir = Path('snapshots')
    if not snapshots_dir.exists():
        snapshots_dir.mkdir(parents=True, exist_ok=True)
        print(f"📁 Создана папка: {snapshots_dir}")
    
    # Корректируем путь к выходному файлу
    final_output_file = snapshots_dir / Path(output_file).name
    
    print(f"📁 Создание снапшота папки: {source_dir}")
    print(f"📄 Выходной файл: {final_output_file}")
    
    markdown = f"# Снапшот папки {source_name}\n\n"
    markdown += f"*Создано: {datetime.now().strftime('%d.%m.%Y %H:%M:%S')}*\n\n"
    
    # Добавляем дерево структуры
    print('🌳 Генерация дерева файлов...')
    markdown += '## Структура файлов\n\n```\n'
    markdown += f"{source_name}/\n"
    
    try:
        items = []
        for item in source_path.iterdir():
            if should_exclude(item.name, item.is_dir()):
                continue
            if item.is_file() and not should_include_file(item.name):
                continue
            items.append(item)
        
        items.sort(key=lambda x: (x.is_file(), x.name.lower()))
        
        for i, item in enumerate(items):
            is_last = i == len(items) - 1
            
            if item.is_dir():
                markdown += create_file_tree(item, '', is_last, 1)
            else:
                connector = '└── ' if is_last else '├── '
                markdown += f"{connector}{item.name}\n"
                
    except Exception as e:
        print(f"❌ Ошибка при создании дерева: {e}")
    
    markdown += '```\n\n---\n\n'
    
    # Собираем все файлы
    print('📋 Сбор файлов...')
    files = collect_files(source_path)
    print(f"📊 Найдено файлов: {len(files)}")
    
    if not files:
        print('⚠️  Файлы для включения в снапшот не найдены')
        return
    
    # Добавляем содержимое файлов
    markdown += '## Содержимое файлов\n\n'
    
    files.sort(key=lambda x: str(x))
    
    for i, file_path in enumerate(files):
        relative_path = file_path.relative_to(source_path)
        normalized_path = str(relative_path).replace('\\', '/')  # Для Windows
        
        print(f"📄 Обработка [{i + 1}/{len(files)}]: {normalized_path}")
        
        try:
            content = file_path.read_text(encoding='utf-8', errors='ignore')
            language = get_language_by_extension(file_path.name)
            
            markdown += f"### {normalized_path}\n"
            markdown += f"```{language}\n"
            markdown += content
            markdown += '\n```\n\n'
            
        except Exception as e:
            print(f"⚠️  Не удалось прочитать файл {file_path}: {e}")
            markdown += f"### {normalized_path}\n"
            markdown += f"*Ошибка чтения файла: {e}*\n\n"
    
    # Записываем результат
    try:
        final_output_file.write_text(markdown, encoding='utf-8')
        print(f"✅ Снапшот создан: {final_output_file}")
        
        size_kb = final_output_file.stat().st_size // 1024
        print(f"📊 Размер файла: {size_kb} KB")
        
    except Exception as e:
        print(f"❌ Ошибка записи файла: {e}")
        sys.exit(1)

def main():
    """Главная функция"""
    parser = argparse.ArgumentParser(
        description='📸 Создание снапшота папки с кодом',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=f"""
Примеры:
  python scripts/create-snapshot.py src src-snapshot.md
  python scripts/create-snapshot.py components components-snapshot.md  
  python scripts/create-snapshot.py . full-project-snapshot.md

Настройки:
  Поддерживаемые расширения: {', '.join(sorted(CONFIG['include_extensions']))}
  Исключаемые папки: {', '.join(sorted(CONFIG['exclude_patterns']))}
  Максимальный размер файла: {CONFIG['max_file_size'] // 1024} KB
  Максимальная глубина: {CONFIG['max_depth']} уровней
        """
    )
    
    parser.add_argument('folder', nargs='?', default='src',
                       help='Папка для создания снапшота (по умолчанию: src)')
    parser.add_argument('output', nargs='?', 
                       help='Выходной файл (по умолчанию: [папка]-snapshot.md)')
    
    args = parser.parse_args()
    
    if not args.output:
        folder_name = Path(args.folder).name
        args.output = f"{folder_name}-snapshot.md"
    
    print('📸 Создание снапшота...')
    create_snapshot(args.folder, args.output)

if __name__ == '__main__':
    main()
